From 171af329d23ef845fbd02e865669c9db8c310500 Mon Sep 17 00:00:00 2001
From: Shikha Panwar <shikhapanwar@google.com>
Date: Thu, 24 Mar 2022 08:54:43 +0000
Subject: [PATCH] Virtualization service to listen for tombstones at VSock port
 from VMs

tombstone_transmit binary running inside the VMs are the Vsock
clients. virtualization_service receives (guest)tomstones from them, requests
for (host)tombstone file to write on -> writes on the file & then
notify tombstoned of the completion (via tombstoned_client)

Test:m succeeds, tested the whole topic by manually crashing random
proceses inside VM (ran them via microdroid demo)

Bug: 202153827
Change-Id: Ifcaa5da968ef39fdd05612d3e0baca4fd1c5eaf1
---
 microdroid/Android.bp                         |  3 +-
 microdroid/bootconfig.common                  |  1 +
 microdroid/init.rc                            |  5 ++
 virtualizationservice/Android.bp              |  2 +
 .../IVirtualMachineService.aidl               |  6 ++
 virtualizationservice/src/aidl.rs             | 68 ++++++++++++++++++-
 6 files changed, 82 insertions(+), 3 deletions(-)

diff --git a/microdroid/Android.bp b/microdroid/Android.bp
index 4631cf8..7ab0978 100644
--- a/microdroid/Android.bp
+++ b/microdroid/Android.bp
@@ -60,7 +60,7 @@ android_system_image {
         "logcat",
         "logd",
         "secilc",
-
+        
         // "com.android.adbd" requires these,
         "libadbd_auth",
         "libadbd_fs",
@@ -77,6 +77,7 @@ android_system_image {
         "linkerconfig",
         "servicemanager.microdroid",
         "tombstoned",
+        "tombstone_transmit",
         "cgroups.json",
         "public.libraries.android.txt",
 
diff --git a/microdroid/bootconfig.common b/microdroid/bootconfig.common
index eda95a2..71cbdae 100644
--- a/microdroid/bootconfig.common
+++ b/microdroid/bootconfig.common
@@ -1,2 +1,3 @@
+androidboot.selinux = permissive
 androidboot.first_stage_console = 1
 androidboot.hardware = microdroid
diff --git a/microdroid/init.rc b/microdroid/init.rc
index f6d5092..6d6a6bf 100644
--- a/microdroid/init.rc
+++ b/microdroid/init.rc
@@ -167,12 +167,17 @@ on post-fs-data
     mkdir /data/vendor/tombstones 0771 root root
 
     start tombstoned
+    start tombstone_transmit
 
     # For security reasons, /data/local/tmp should always be empty.
     # Do not place files or directories in /data/local/tmp
     mkdir /data/local 0751 root root
     mkdir /data/local/tmp 0771 shell shell
 
+service tombstone_transmit /system/bin/tombstone_transmit -cid 2 -port 2000
+    user root
+    group system
+
 service apexd-vm /system/bin/apexd --vm
     user root
     group system
diff --git a/virtualizationservice/Android.bp b/virtualizationservice/Android.bp
index 9b2b740..a6560fe 100644
--- a/virtualizationservice/Android.bp
+++ b/virtualizationservice/Android.bp
@@ -45,6 +45,7 @@ rust_defaults {
         "libserde_xml_rs",
         "libshared_child",
         "libstatslog_virtualization_rust",
+        "libtombstoned_client_bindgen",
         "libvmconfig",
         "libzip",
         "libvsock",
@@ -54,6 +55,7 @@ rust_defaults {
     shared_libs: [
         "libbinder_rpc_unstable",
         "libselinux",
+        "libtombstoned_client",
     ],
 }
 
diff --git a/virtualizationservice/aidl/android/system/virtualmachineservice/IVirtualMachineService.aidl b/virtualizationservice/aidl/android/system/virtualmachineservice/IVirtualMachineService.aidl
index 1a16f2a..dff5d46 100644
--- a/virtualizationservice/aidl/android/system/virtualmachineservice/IVirtualMachineService.aidl
+++ b/virtualizationservice/aidl/android/system/virtualmachineservice/IVirtualMachineService.aidl
@@ -29,6 +29,12 @@ interface IVirtualMachineService {
      */
     const int VM_BINDER_SERVICE_PORT = 5000;
 
+    /**
+     * Port number that VirtualMachineService listens on connections from the guest VMs for the
+     * tombtones
+     */
+    const int VM_TOMBSTONES_SERVICE_PORT = 2000;
+
     /**
      * Notifies that the payload has started.
      */
diff --git a/virtualizationservice/src/aidl.rs b/virtualizationservice/src/aidl.rs
index d9825dc..a035ebe 100644
--- a/virtualizationservice/src/aidl.rs
+++ b/virtualizationservice/src/aidl.rs
@@ -42,7 +42,7 @@ use android_system_virtualizationservice::binder::{
 use android_system_virtualmachineservice::aidl::android::system::virtualmachineservice::{
     IVirtualMachineService::{
         BnVirtualMachineService, IVirtualMachineService, VM_BINDER_SERVICE_PORT,
-        VM_STREAM_SERVICE_PORT,
+        VM_STREAM_SERVICE_PORT, VM_TOMBSTONES_SERVICE_PORT,
     },
 };
 use anyhow::{anyhow, bail, Context, Result};
@@ -57,13 +57,14 @@ use statslog_virtualization_rust::vm_creation_requested::{stats_write, Hyperviso
 use std::convert::TryInto;
 use std::ffi::CStr;
 use std::fs::{create_dir, File, OpenOptions};
-use std::io::{Error, ErrorKind, Write};
+use std::io::{Error, ErrorKind, Write, Read};
 use std::num::NonZeroU32;
 use std::os::raw;
 use std::os::unix::io::{FromRawFd, IntoRawFd};
 use std::path::{Path, PathBuf};
 use std::ptr::null_mut;
 use std::sync::{Arc, Mutex, Weak};
+use tombstoned_client::*;
 use vmconfig::VmConfig;
 use vsock::{SockAddr, VsockListener, VsockStream};
 use zip::ZipArchive;
@@ -86,6 +87,7 @@ const ANDROID_VM_INSTANCE_MAGIC: &str = "Android-VM-instance";
 /// Version of the instance image format
 const ANDROID_VM_INSTANCE_VERSION: u16 = 1;
 
+const CHUNK_RECV_MAX_LEN:usize = 1024;
 /// Implementation of `IVirtualizationService`, the entry point of the AIDL service.
 #[derive(Debug, Default)]
 pub struct VirtualizationService {
@@ -371,6 +373,62 @@ impl IVirtualizationService for VirtualizationService {
     }
 }
 
+fn handle_stream_connection_tombstoned()-> Result<()> {
+    let listener = VsockListener::bind_with_cid_port(VMADDR_CID_HOST, VM_TOMBSTONES_SERVICE_PORT as u32)?;    
+    for incoming_stream in listener.incoming() {
+        let mut incoming_stream = match incoming_stream {
+            Err(e) => {
+                warn!("invalid incoming connection: {}", e);
+                continue;
+            }
+            Ok(s) => s,
+        };
+        std::thread::spawn(move || {
+            if let Err(e) = handle_tombstone(&mut incoming_stream) {
+                error!("Failed to write tombstone- {:?}", e);
+            }
+        });
+    }
+    Ok(())
+}
+
+fn handle_tombstone(stream: &mut VsockStream)-> Result<()> {
+    if let Ok(SockAddr::Vsock(addr)) = stream.peer_addr() {
+        info!("Vsock Stream connected to cid={} for tombstones", addr.cid());
+    }
+    unsafe {
+        let mut tombstoned_socket = android_base_unique_fd{ fd_: -1 };
+        let rp_tombstoned_socket = &mut tombstoned_socket as *mut android_base_unique_fd;
+
+        let mut tombstone_fd = android_base_unique_fd{ fd_: -1 };
+        let rp_tombstone_fd = &mut tombstone_fd as *mut android_base_unique_fd;
+
+        let g_tombstoned_connected = tombstoned_connect1(std::process::id() as i32, rp_tombstoned_socket, 
+                rp_tombstone_fd, DebuggerdDumpType_kDebuggerdTombstone);
+        if !g_tombstoned_connected {
+            return Err(anyhow!("Could not connect to tombstoned"));
+        }
+        let mut file_tombstone = File::from_raw_fd((*rp_tombstone_fd).fd_);
+        let mut num_bytes_read =  0;
+        loop {
+            let mut chunk_recv = [0; CHUNK_RECV_MAX_LEN];
+            let n = stream.read(&mut chunk_recv)?;
+            if n == 0 {
+                break;
+            }
+            num_bytes_read += n;
+            file_tombstone.write_all(&chunk_recv[0..n])?;
+        }
+        info!("Received {} bytes from guest & wrote to tombstone file", num_bytes_read);
+        // Need to call into_raw_fd to release the ownership of fd.
+        let _ = file_tombstone.into_raw_fd();
+        if !tombstoned_notify_completion(tombstoned_socket.fd_) {
+            warn! ("failed to notify tombstoned of completion");
+        }
+    }
+    Ok(())
+}
+
 impl VirtualizationService {
     pub fn init() -> VirtualizationService {
         let service = VirtualizationService::default();
@@ -381,6 +439,12 @@ impl VirtualizationService {
             handle_stream_connection_from_vm(state).unwrap();
         });
 
+        std::thread::spawn(move || {
+            if let Err(e) = handle_stream_connection_tombstoned() {
+                warn!("Error receiving tombstone from guest or writing them. Error: {}", e);
+            }
+        });
+
         // binder server for vm
         let mut state = service.state.clone(); // reference to state (not the state itself) is copied
         std::thread::spawn(move || {
-- 
2.35.1.894.gb6a874cedc-goog

